src/utils/point_locator_nanoflann.C

```diff
@@ -0,0 +1,427 @@
+  // stored in _b so that _b[0] is the first result to check, _b[1]
+  // is second, etc.
+  _b.resize(result_set.size());
+  std::iota(_b.begin(), _b.end(), 0);
+  std::sort(_b.begin(), _b.end(), comp);
+
+  // Note: we use result_set.size() here rather than
+  // _num_results, in case the result returns fewer than
+  // _num_results results!
+  for (auto r : make_range(result_set.size()))
+    {
+      // Translate the Nanoflann index, which is from [0..n_points),
+      // into the corresponding Elem id from the mesh. Note that we
+      // use index _b[r] instead of r, since we want to process the
+      // Elems in order based on Elem id.
+      auto nanoflann_index = _ret_index[_b[r]];
+      const Elem * candidate_elem = (*_elems)[nanoflann_index];
+
+      // Debugging: print the results
+      // libMesh::err << "Centroid/Elem id = " << candidate_elem->id()
+      //              << ", dist^2 = " << _out_dist_sqr[r]
+      //              << std::endl;
+
+      // Before we even check whether the candidate Elem actually
+      // contains the Point, we may need to check whether the
+      // candidate Elem is from an allowed subdomain.  If the
+      // candidate Elem is not from an allowed subdomain, we continue
+      // to the next one.
+      if (allowed_subdomains && !allowed_subdomains->count(candidate_elem->subdomain_id()))
+        {
+          // Debugging
+          // libMesh::err << "Elem " << candidate_elem->id() << " was not from an allowed subdomain, continuing search." << std::endl;
+          continue;
+        }
+
+      // If we made it here, then the candidate Elem is from an
+      // allowed subdomain, so let's next check whether it contains
+      // the point. If the user set a custom tolerance, then we
+      // actually check close_to_point() rather than contains_point(),
+      // since this latter function warns about using non-default
+      // tolerances, but otherwise does the same test.
+      bool inside = _use_contains_point_tol ?
+        candidate_elem->close_to_point(p, _contains_point_tol) :
+        candidate_elem->contains_point(p);
+
+      // Increment the number of elements checked
+      n_elems_checked++;
+
+      // If the point is inside an Elem from an allowed subdomain, we are done.
+      if (inside)
+        {
+          // Debugging: report the number of Elems checked
+          // libMesh::err << "Checked " << n_elems_checked << " nearby Elems before finding a containing Elem." << std::endl;
+
+          found_elem = candidate_elem;
+          break;
+        }
+    } // end for(r)
+
+  // If we made it here, then at least one of the following happened:
+  // .) The search Point was not in the BoundingBox of local Elems.
+  // .) All the candidate elements were from non-allowed subdomains.
+  // .) The Point was not inside _any_ of the _num_results candidate Elems.
+  // Thus, if we are not in _out_of_mesh_mode, throw an error,
```

Comment:
Oh, no.

This seems like it can give us false negatives, doesn't it?

I was just talking on Friday with someone whose use case was "a 2D manifold, intersecting a 3D manifold, possibly of very different element sizes".  What happens in that case if we're looking (via subdomain_id restriction) for a 3D element, but the nearest `_num_results` elements all happen to be 2D because there's so many more of them?

In fact, we might not even have to get that wild.  The default `_num_results` is only 8?  If I have 8 cube hexes meeting at a vertex V, and I refine one of them, call it R, three times (refining it, then all its children, then all its children's children), and I let the default face-neighbor level-one rule refine part of its neighbors twice and part of its neighbors' neighbors once ... then there's one remaining unrefined element U, and inside that element near V I'd expect at least *eighty* elements to have centroids closer to that corner of U than the centroid of U is: one among each of the three once-refined neighbors' children, 8 among each of the three twice-refined neighbors' grandchildren, and 64+ of R's great-grandchildren.

We've got to have some better solution here than "report a false negative", don't we?  It can't be a *common* problem or we'd have triggered it in CI, so you'd hope we could afford an expensive backup plan knowing that the expense will be amortized by the 0.lots-of-zeros-1 % of times we actually need it ... but we'll *also* end up hitting the backup plan in cases where the user really does query for an out-of-mesh element (which can happen a lot on distributed meshes; even System::point_value() does it!) and that's too expensive.

Reply:
And I'm already stymied by this problem for the "just return one element" overload ...  Figuring out a cheap backup plan for the "return the entire set of elements and be sure not to miss any" overload might be one step harder still.

Reply:
Yeah, that is definitely an issue... you can an elaborate "backup" plan, but the problem is that you are going to hit it a majority of the time in the `DistributedMesh` case, since assuming the points are uniformly distributed, you are only going to find the Point about 1 in n_procs searches.

For the case that you describe, I would suggest that we store the centroids/make the KD-Trees by subdomain as well. Then we'd only search the KD-Tree(s) of the allowed subdomain(s) when that option is active. It makes the bookkeeping a bit more complicated, but I think it would be possible.

Reply:
I'm thinking of more pathological cases as fast as I'm thinking of possible solutions, sadly.  And most of them are more like the AMR case than like the 3D+2D case - with only one subdomain, no subdomain-based solution will fix them.

We can't even help things by using an initial semilocal bounding box check in the DistributedMesh case, can we?  It won't matter if only ~3 processors are hitting the hypothetical backup plan rather than ~N-3 processors, because in most codes the other processors are still going to end up twiddling their thumbs waiting for the slow 3.

Reply:
Gah, I don't even need AMR to make the vertices-tree idea fail.  You can have an arbitrary number of triangles meet at one vertex, so unless you're looking at an arbitrary number of results you can't be sure that the triangle you're actually contained in is among them.

Reply:
> I can combine AMR (to get a hanging node) with a refined boundary layer (to put an arbitrary number of thin elements on the refined side of that node) to get an arbitrary number of vertices closer to a point than any of the point's containing element's own vertices are.

In this scenario, are you not assuming the level one rule is active? Sorry, just trying to draw a picture...

Reply:
Level-one rule active, so in 2D you only get one hanging node ... but if you have a very-fine-in-the-normal-direction mesh on one side of the hanging node and a more isotropic mesh on the other side (Yes, you'd normally want a smoother transition, but don't laugh!  I recall seeing pictures of just such meshes hanging on a wall in WRW!) then you can have lots of boundary-layer nodes that are closer than the isotropic element's centroid.

And no need to apologize for anything; I feel like a jackass here.  I kinda had this sort of scenario in the back of my head when I first looked at this PR, but I didn't really think about it and I just assumed that it was the sort of thing that would be easy to fix if it really was a problem.

Reply:
Would the old `PointLocator` not also have issues with the different cases you've mentioned? 

I'm OK with the Nanoflann PL not becoming the default for everyone, but from our perspective, we are still interested in using it. (We don't use AMR or DistributedMesh, so that rules out some of the tricky cases straight out of the box.) If I make the Nanoflann PointLocator a non-default configure time option, would that be acceptable? The issue with mixed-dimension subdomains that you mentioned might actually be relevant to us... but I won't know until I have a chance to test it further.

Reply:
Hmm... I guess the "TREE" PointLocatorType might have analogous problems?  That gets translated into Tree::NODES, which just looks at the element's nodes and not its bounding box.  Maybe that would still work in the single-manifold case, but at least in the overlapping-manifold case I could definitely imagine large 3D elements whose nodes are too "sparse" relative to the octree generated to capture a bunch of finer 2D elements cutting through the coarser volume mesh.

But we always (well, for a decade plus at least) defaulted to TREE_ELEMENTS/Trees::ELEMENTS, which uses bounding box intersections to avoid this problem.  Doesn't matter how far your element's center or its nodes is from an octree cell, if there's some intersection between the element's bounding box and the cell then when we hit that cell (assuming it's a leaf) we check that element.

> We don't use AMR or DistributedMesh, so that rules out some of the tricky cases straight out of the box.

DistributedMesh doesn't actually create any failure cases that ReplicatedMesh doesn't, it just makes them much more expensive to imagine fixing.

Avoiding AMR would definitely keep this from being a common concern, but it wouldn't stop it from being a theoretical concern.  Any hanging node quad mesh could be turned into a conforming triangle mesh with nearly the exact same problem.

> If I make the Nanoflann PointLocator a non-default configure time option, would that be acceptable?

If you plaster the doxygen with warnings, too, then definitely!  If I didn't believe in people's inalienable right to have enough rope to hang themselves with, I wouldn't even be programming in C++.  You might never run into these issues in your problems, or you might run into them but we might think of a real solution after we've had some time to stew about it.

Or we might decide that we want to try some hybrid in the future.  If the Nanoflann PointLocator is sufficiently faster and works 99% of the time, I could imagine using the PointLocatorTree as our backup plan.  Setup time and memory usage would be worse than either alone (say, any idea if the Nanoflann KDTree size is much better or worse than our Tree?) but if the result was just as robust as PointLocatorTree but we still got 99% of the lookup speedup of the Nanoflann locator then that might be worth it, at least to anybody using a distributed mesh or a high-memory machine (more and more people in both cases).

> The issue with mixed-dimension subdomains that you mentioned might actually be relevant to us...

One other possibly-Akselos-relevant case I just thought of - do you guys still sometimes use those "spider" meshes with a ton of EDGE2 elements connected to a common node to create long-distance algebraic coupling?  I assume you never do any point lookups on those edges, but if you do then I'd say expect them to fail, approximately a ```_num_results/num_edges``` fraction of the time.

Reply:
> One other possibly-Akselos-relevant case I just thought of - do you guys still sometimes use those "spider" meshes with a ton of EDGE2 elements connected to a common node to create long-distance algebraic coupling? I assume you never do any point lookups on those edges, but if you do then I'd say expect them to fail, approximately a _num_results/num_edges fraction of the time.

One of the motivations behind developing this new PointLocator was to handle such cases, yes. With the current `PointLocatorTree`, we have to use the `MeshBase::set_count_lower_dim_elems_in_point_locator(bool count_lower_dim_elems)` "hack" to avoid hitting an infinite loop in the spider node case. 

The EDGE2 elements don't typically overlap for us in the spider node case... you can imagine them as being like spokes on a wheel on a circular geometry, for example. So I don't foresee issues with false negatives in that configuration. I guess we'll know if that's the case after we do some more extensive internal testing.